contrib/hualloc/hu_alloc.h

```diff
@@ -0,0 +1,1615 @@
+{
+    // no special action required
+    (void)p;
+    (void)sz;
+}
+static void OsReclaim(void *p, yint sz)
+{
+    //madvise(segment, SEGMENT_SIZE, MADV_DONTNEED); // super slow for no apparent reason, use MAP_FIXED to map new memory over non needed
+    void *res = mmap(p, sz, PROT_READ | PROT_WRITE, MAP_FIXED | MAP_PRIVATE | MAP_ANON | MAP_NORESERVE, -1, 0);
+    if (res != p) {
+        abort(); // MAP_FIXED failed
+    }
+#ifdef MADV_HUGEPAGE
+    madvise(res, sz, MADV_HUGEPAGE);
+#endif
+}
+static void* OsAlloc(yint *pSz)
+{
+    yint &sz = *pSz;
+    bool isHuge = false;
+    if (sz > 5 * 1000 * 1000) {
+        // huge alloc
+        const yint hugePageSize = 1ll << 21;
+        sz = (sz + hugePageSize - 1) & ~(hugePageSize - 1);
+        isHuge = true;
+    } else {
+        // regular alloc
+        sz = (sz + PAGE_SIZE - 1) & ~(PAGE_SIZE - 1);
+    }
+    void *ptr = mmap(0, sz, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
+    if (isHuge) {
+#ifdef MADV_HUGEPAGE
+        madvise(ptr, sz, MADV_HUGEPAGE);
+#endif
+    }
+    Y_ASSERT(ptr);
+    return ptr;
+}
+static void OsFree(void *p, yint sz)
+{
+    munmap(p, sz);
+}
+#endif
+
+static void* OsAlloc(yint szArg)
+{
+    yint sz = szArg;
+    return OsAlloc(&sz);
+}
+
+// smallest power of 2 greater or equal to val
+static yint CalcLog2(yint val)
+{
+    if (val <= 1) {
+        return 0;
+    }
+    unsigned long xx;
+    _BitScanReverse64(&xx, val - 1);
+    return xx + 1;
+}
+
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+static std::atomic<ui64> AllocatorIsInitialized;
+static void hu_init();
+static void hu_check_init()
+{
+    if (AllocatorIsInitialized != 1) {
+        hu_init();
+    }
+}
+
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+const yint GLOBAL_HASH_SIZE_LN = 20; // 4x+ number of blocks and more to avoid excessive hash lookups
+const yint GLOBAL_HASH_SIZE = 1ll << GLOBAL_HASH_SIZE_LN;
+
+struct TGAEntry
+{
+    std::atomic<ui64> Ptr;
+    yint Size;
+};
+static TGAEntry *GAHash;
+
+
+static void GAInit()
+{
+    GAHash = (TGAEntry*)OsAlloc(GLOBAL_HASH_SIZE * sizeof(TGAEntry));
+}
+
+
+static yint GAHashPtr(const void *ptr)
+{
+    ui64 x = (const char*)(ptr) - (const char*)nullptr;
+    ui64 res = (0x4847badfea31337ull * x) >> (64 - GLOBAL_HASH_SIZE_LN);
+    Y_ASSERT(res < GLOBAL_HASH_SIZE);
+    return res;
+}
+
+
+NOINLINE static void* GAAlloc(yint szArg)
+{
+    if (szArg == 0) {
+        return 0;
+    }
+    hu_check_init();
+    yint sz = szArg;
+    void *ptr = OsAlloc(&sz);
+    if (ptr == 0) {
+        abort(); // out of memory
+    }
+    yint k = GAHashPtr(ptr);
+    for (yint kStart = k;;) {
+        TGAEntry &ga = GAHash[k];
+        ui64 oldPtr = ga.Ptr.load();
+        if (oldPtr <= 1) {
+            ui64 newPtr = (ui64)ptr;
+            if (ga.Ptr.compare_exchange_weak(oldPtr, newPtr)) {
```

Comment:
compare_exchange_weak looked suspicious, so I checked [the standard](https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange):

> The weak forms (1-2) of the functions are allowed to fail spuriously, that is, act as if *this != expected even if they are equal. When a compare-and-exchange is in a loop, the weak version will yield better performance on some platforms.

This means that the value may be 0 and still remain 0, but we will go to the next slot.

Reply:
Looks like a bug indeed. Does not reproduce on x64 though, weak and strong are same lock cmpxchg for x64.